spec/app/ics-021-nft-transfer/README.md

```diff
@@ -0,0 +1,360 @@
+---
+ics: 21
+title: Non-Fungible Token Transfer
+stage: draft
+category: IBC/APP
+requires: 25, 26
+kind: instantiation
+author: Christopher Goes <cwgoes@interchain.berlin>, Haifeng Xi <haifeng@bianjie.ai>
+created: 2021-11-10 
+modified: 2021-11-10
+---
+
+## Synopsis
+
+This standard document specifies packet data structure, state machine handling logic, and encoding details for the transfer of non-fungible tokens over an IBC channel between two modules on separate chains. The state machine logic presented allows for safe multi-chain denomination handling with permissionless channel opening. This logic constitutes a "non-fungible token transfer bridge module", interfacing between the IBC routing module and an existing asset tracking module on the host state machine.
+
+### Motivation
+
+Users of a set of chains connected over the IBC protocol might wish to utilise an asset issued on one chain on another chain, perhaps to make use of additional features such as exchange or privacy protection, while retaining non-fungibility with the original asset on the issuing chain. This application-layer standard describes a protocol for transferring non-fungible tokens between chains connected with IBC which preserves asset non-fungibility, preserves asset ownership, limits the impact of Byzantine faults, and requires no additional permissioning.
+
+### Definitions
+
+The IBC handler interface & IBC routing module interface are as defined in [ICS 25](../../core/ics-025-handler-interface) and [ICS 26](../../core/ics-026-routing-module), respectively.
+
+### Desired Properties
+
+- Preservation of uniqueness (two-way peg).
+- Preservation of total supply (maintained on a single source chain & module).
+- Permissionless token transfers, no need to whitelist connections, modules, or denominations.
+- Symmetric (all chains implement the same logic, no in-protocol differentiation of hubs & zones).
+- Fault containment: prevents Byzantine-inflation of tokens originating on chain `A`, as a result of chain `B`'s Byzantine behaviour (though any users who sent tokens to chain `B` may be at risk).
+
+## Technical Specification
+
+### Data Structures
+
+Only one packet data type is required: `NonFungibleTokenPacketData`, which specifies the denomination, amount, sending account, and receiving account.
+
+```typescript
+interface NonFungibleTokenPacketData {
+  classId: string
+  classUri: string
```

Comment:
do we need to send URI?

Reply:
Class URI is optional but will be very useful for cross-chain interoperability with NFT marketplace like OpenSea where class/collection metadata can be configured for better user experience.

Reply:
> Class URI is optional but will be very useful for cross-chain interoperability with NFT marketplace like OpenSea where class/collection metadata can be configured for better user experience.

https://docs.opensea.io/docs/contract-level-metadata

Reply:
Makes sense, could you add a note / comment about it in the file, please?

Reply:
comment added to the spec

Reply:
In the event that a token URI is updated on the origin chain after the NFT has been transferred to another chain, will it cause a problem when the NFT is transferred back since the IBC packet's TokenURI will be different from the origin? Or is it not possible to update the token URI on the origin chain when it is locked in escrow?

Reply:
Once a token is transferred out, the original token has been locked in the escrow account at source, and the correct business logic at source, I believe, should be that no updates can be made to locked tokens.

Reply:
> the locked token in the escrow account owned by IBC indicates that NFT has a third status as "locked", so should we describe this status in NFT data structure?

NFT status is a local attribute that is tracked by the application layer at source and participating chains. I think we don't need to include it in the transport protocol.

Reply:
> I think the correct logic should be: only tokens that have not been transferred out via IBC can have their tokenURI updated by owner. 

Makes sense - let's add it to the spec. How we can ensure it in x/nft? Today, there is no functionality to update NFT. Maybe we should add it on the keeper level (not msg server level)? The keeper will check if nft is owned by x/ibc to allow edits. What do you think @haifengxi ? If yes then we need to create a new task.

Reply:
Will update the spec.  However, this is something that can only be enforced by the application layer of participating chains instead of the IBC protocol itself.  `x/nft` is at the "storage layer", we probably cannot enforce it there without incurring a circular dependency between `x/nft` and `x/ibc`.